File: app/src/main/java/com/example/notes/data/CardsSourceImpl.java
package com.example.notes.data;

import android.content.res.Resources;
import android.content.res.TypedArray;

import com.example.notes.R;

import java.util.ArrayList;
import java.util.List;

public class CardsSourceImpl implements CardsSource {

    private List<CardData> dataSource;
    private Resources resources; // ресурсы приложения

    public CardsSourceImpl(Resources resources) {
        dataSource = new ArrayList<>(7);
        this.resources = resources;
    }

    public CardsSourceImpl init() {
        // строки заголовков из ресурсов
        String[] titles = resources.getStringArray(R.array.titles);
        // строки описаний из ресурсов
        String[] descriptions = resources.getStringArray(R.array.descriptions);
        // изображения
        int[] pictures = resources.getIntArray(R.array.pictures);
        // заполнение источника данных
        for (int i = 0; i < descriptions.length; i++) {
            dataSource.add(new CardData(titles[i], descriptions[i], pictures[i], false));
        }
        return this;
    }

    // Механизм вытаскивания идентификаторов картинок
    private int[] getImageArray(){
        TypedArray pictures = resources.obtainTypedArray(R.array.pictures);
        int length = pictures.length();
        int[] answer = new int[length];
        for (int i = 0; i < length; i++) {
            answer[i] = pictures.getResourceId(i, 0);
        }
        return answer;
    }

    @Override
    public CardData getCardData(int position) {
        return dataSource.get(position);
    }

    @Override
    public int size() {
        return dataSource.size();
    }
}

File: app/src/main/java/com/example/notes/data/CardsSourceImpl.kt
package com.example.notes.data

import android.content.res.Resources
import com.example.notes.R
import java.util.*

class CardsSourceImpl(resources: Resources) : CardsSource {
    private val dataSource: MutableList<CardData?>
    private val resources // ресурсы приложения
            : Resources

    override fun init(cardsSourceResponse: CardsSourceResponse?): CardsSource? {
        // строки заголовков из ресурсов
        val titles = resources.getStringArray(R.array.titles)
        // строки описаний из ресурсов
        val descriptions = resources.getStringArray(R.array.descriptions)
        // изображения
        val pictures = imageArray
        // заполнение источника данных
        for (i in descriptions.indices) {
            dataSource.add(
                CardData(
                    titles[i],
                    descriptions[i],
                    pictures[i],
                    false,
                    Calendar.getInstance().time
                )
            )
        }
        cardsSourceResponse?.initialized(this)
        return this
    }

    // Механизм вытаскивания идентификаторов картинок
    private val imageArray: IntArray
        private get() {
            val pictures = resources.obtainTypedArray(R.array.pictures)
            val length = pictures.length()
            val answer = IntArray(length)
            for (i in 0 until length) {
                answer[i] = pictures.getResourceId(i, 0)
            }
            return answer
        }

    override fun getCardData(position: Int): CardData? {
        return dataSource[position]
    }

    override fun size(): Int {
        return dataSource.size
    }

    override fun deleteCardData(position: Int) {
        dataSource.removeAt(position)
    }

    override fun addCardData(cardData: CardData?) {
        dataSource.add(cardData)
    }

    override fun clearCardData() {
        dataSource.clear()
    }

    override fun updateCardData(position: Int, cardData: CardData?) {
        dataSource[position] = cardData
    }

    init {
        dataSource = ArrayList(7)
        this.resources = resources
    }
}